File: eVetCare.Services/Database/MedicalRecord.cs
namespace eVetCare.Services.Database;

public partial class MedicalRecord
{
    public int MedicalRecordId { get; set; }

    public int PetId { get; set; }

    public int AppointmentId { get; set; }

    public DateTime Date { get; set; }

    public string? Notes { get; set; }

    public string? AnalysisProvided { get; set; }

    public bool? IsDeleted { get; set; }

    public bool IsActive { get; set; }

    public virtual Appointment Appointment { get; set; } = null!;

    public virtual ICollection<Diagnosis> Diagnoses { get; set; } = new List<Diagnosis>();

    public virtual ICollection<LabResult> LabResults { get; set; } = new List<LabResult>();

    public virtual Pet Pet { get; set; } = null!;

    public virtual ICollection<Treatment> Treatments { get; set; } = new List<Treatment>();

    public virtual ICollection<Vaccination> Vaccinations { get; set; } = new List<Vaccination>();
}

File: UI/evetcare_admin/lib/models/medical_record.dart
import 'package:json_annotation/json_annotation.dart';
import 'diagnosis.dart';
import 'treatment.dart';
import 'lab_result.dart';
import 'vaccination.dart';

part 'medical_record.g.dart';

@JsonSerializable()
class MedicalRecord {
  final int medicalRecordId;
  final int petId;
  final String petName;
  final int appointmentId;
  final DateTime date;
  final String? notes;
  final String? analysisProvided;
  final List<Diagnosis> diagnoses;
  final List<Treatment> treatments;
  final List<LabResult> labResults;
  final List<Vaccination> vaccinations;

  MedicalRecord({
    required this.medicalRecordId,
    required this.petId,
    required this.petName,
    required this.appointmentId,
    required this.date,
    this.notes,
    this.analysisProvided,
    required this.diagnoses,
    required this.treatments,
    required this.labResults,
    required this.vaccinations,
  });

  factory MedicalRecord.fromJson(Map<String, dynamic> json) => MedicalRecord(
    medicalRecordId: json['medicalRecordId'] as int,
    petId: json['petId'] as int,
    petName: json['petName'] as String,
    appointmentId: json['appointmentId'] as int,
    date: DateTime.parse(json['date'] as String),
    notes: json['notes'] as String?,
    analysisProvided: json['analysisProvided'] as String?,
    diagnoses:
        (json['diagnoses'] as List<dynamic>?)
            ?.map((e) => Diagnosis.fromJson(e as Map<String, dynamic>))
            .toList() ??
        [],
    treatments:
        (json['treatments'] as List<dynamic>?)
            ?.map((e) => Treatment.fromJson(e as Map<String, dynamic>))
            .toList() ??
        [],
    labResults:
        (json['labResults'] as List<dynamic>?)
            ?.map((e) => LabResult.fromJson(e as Map<String, dynamic>))
            .toList() ??
        [],
    vaccinations:
        (json['vaccinations'] as List<dynamic>?)
            ?.map((e) => Vaccination.fromJson(e as Map<String, dynamic>))
            .toList() ??
        [],
  );

  Map<String, dynamic> toJson() => {
    'medicalRecordId': medicalRecordId,
    'petId': petId,
    'petName': petName,
    'appointmentId': appointmentId,
    'date': date.toIso8601String(),
    'notes': notes,
    'analysisProvided': analysisProvided,
    'diagnoses': diagnoses.map((d) => d.toJson()).toList(),
    'treatments': treatments.map((t) => t.toJson()).toList(),
    'labResults': labResults.map((l) => l.toJson()).toList(),
    'vaccinations': vaccinations.map((v) => v.toJson()).toList(),
  };
}
